### DISCUSSION TITLE
Type guard for union types

### DISCUSSION THREAD

**Discussion by zolomatok** [DanielXMoore/Civet#303] [Ideas]
It would be great if we could use _is_ or _as_ something similar to disambiguate between types in a union type, especially if they are not primitives (so when _typeof_ doesn't help). Eg:

```
interface Something
    someVar: string

interface SomeOtherThing
    someOtherVar: string

function test(s: Something | SomeOtherThing)
     if s is Something
          console.log s.someVar
     else if s is SomeOtherThing
          console.log s.someOtherVar
```

**Comment by zolomatok** [comment]
As far as I know this kind of thing is only possible with [user-defined type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates), which are a chore.

**Reply by edemaine** [reply]
If I understand it correctly, I think [DeepKit](https://deepkit.io/) offers one approach to doing this.

**Reply by zolomatok** [reply]
Thanks for the suggestion @edamaine!
DeepKit seems like a huge framework though, I wouldn't use anything that sprawling.

**Comment by edemaine** [comment]
Related, I'm really looking forward to when I can use this in pattern matching:

```ts
interface point
  x: number
  y: string
switch v
  <? point
    ...
---
interface point {
  x: number
  y: string
}
if (typeof v === "object" && v != null && typeof(v.x) === "number" && typeof(v.y) === "string") {
  ...
}
```

Before we understand type references though (looking up `point` is tricky), we could aim for something like this (based on one of the #126 proposals):

```ts
switch v
  {x:: number, y:: string}
```

(originally posted https://github.com/DanielXMoore/Civet/pull/296#issuecomment-1407065657)

**Reply by zolomatok** [reply]
```
switch v
  {x:: number, y:: string}
```
Would checking the prototype allow for checking if it conforms to an interface?
Or would this only be for primitives?

Also if we're making `instanceof` work in pattern matching ...
```
interface point
  x: number
  y: string
switch v
  <? point
    ...
```
... I guess then this would make `if x <? point` valid too?
Though I kinda like `if x is point` more, if that's possible.


I guess we could also just use `discriminated unions` for working with `interfaces` ...
```
interface Cow
    type: 'cow'
    giveMilk : =>

interface Dog
    type: 'dog'
    bark : =>

function doSomething(animal: Cow | Dog)
    switch animal.type
        when 'cow' then animal.giveMilk()
        when 'dog' then animal.bark()
```
... but that would be best done with [enums](https://github.com/DanielXMoore/Civet/discussions/227) for the `type` field for type safety.
